controllers/apps/transformer_cluster_deletion.go

```diff
@@ -76,12 +76,18 @@ func (t *clusterDeletionTransformer) Transform(ctx graph.TransformContext, dag *
 	transCtx.EventRecorder.Eventf(cluster, corev1.EventTypeNormal, constant.ReasonDeletingCR, "Deleting %s: %s",
 		strings.ToLower(cluster.GetObjectKind().GroupVersionKind().Kind), cluster.GetName())
 
-	// delete components in the order that topology defined.
-	if err := deleteCompsInOrder4Terminate(transCtx, dag); err != nil {
+	// firstly, delete components in the order that topology defined.
+	deleteCompSet, err := deleteCompsInOrder4Terminate(transCtx, dag)
+	if err != nil {
 		return err
 	}
+	if len(deleteCompSet) > 0 {
+		// wait for the components to be deleted to trigger the next reconcile
+		transCtx.Logger.Info(fmt.Sprintf("wait for the components to be deleted: %v", deleteCompSet))
```

Comment:
If no deletion dependency is set, then this check is skipped. Components and other resources are still deleted in parallel. The desired effect you're expecting (resources like serviceaccounts and rolebindings not being deleted when the component's preTerminate is called) may not be achievable.